Start integrating direction into 'create RTCRtpTransceiver' algorithm

adam-be opened the pull request:
Fixes #1206

Work in progress

stefhak commented:
@jan-ivar is this going in the right direction?

adam-be commented:
Patched 'create an RTCRtpTransceiver' algorithm. Updated addTrack, addTransceiver and 'set an RTCDescription' steps.

jan-ivar commented on webrtc.html:
```diff
@@ -1243,6 +1243,14 @@
+                      <li>
+                        <p>If <var>description</var> is of type "answer", then
+                        set <var>transceiver</var>'s <code><a data-for=
+                        "RTCRtpTransceiver">currentDirection</a></code> value
```
This is setting a read-only property. See https://github.com/w3c/webrtc-pc/issues/1204.

adam-be replied:
True. I'll have a look at it.

adam-be commented:
I've modified this PR to define and use internal slots for direction and currentDirection. I've also updated the other text to use these slots so that fixes #1204. Please have a look @jan-ivar.

adam-be commented:
Ping @jan-ivar

jan-ivar commented on webrtc.html:
```diff
@@ -5208,6 +5238,12 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
+                  <p>If the dictionary argument is present, let
+                  <var>direction</var> be the value of the
+                  <code>direction</code> member. Otherwise let
+                  <var>direction</var> be <code>sendrecv</code>.</p>
```
FWIW all other uses of "If the dictionary argument is present," are redundant as well.

adam-be replied:
Let's fix all these in #1369

adam-be commented on webrtc.html:
```diff
@@ -1243,6 +1243,14 @@
                         "RTCRtpTransceiver">mid</a></code> value to the mid of
                         the corresponding media description.</p>
                       </li>
+                      <li>
```
> 7. If description is set as a local description, then run the following steps for each media description in description that is not yet associated with an RTCRtpTransceiver object:

This has to do with the mid. If a transceiver has beed assigned the mid of a media description, they are associated.

> 1. Let transceiver be the RTCRtpTransceiver used to create the media description.

The media description was created from data in the transceiver. They might not be associated though.

Clarifications to this part should be done in a separate PR I think.

adam-be commented on webrtc.html:
```diff
@@ -1243,6 +1243,14 @@
                         the corresponding media description.</p>
                       </li>
+                      <li>
+                        <p>If <var>description</var> is of type "answer", then
```
I don't know what to do for `"pranswer"` in this case.

adam-be commented on webrtc.html:
```diff
@@ -1243,6 +1243,14 @@
+                      <li>
+                        <p>If <var>description</var> is of type "answer", then
+                        set <var>transceiver</var>'s [[<a href=
+                        "#curr-direction-slot">CurrentDirection</a>]] slot
```
Thanks for making an issue. This has been brought up at the editor's meeting as well.

adam-be commented on webrtc.html:
```diff
@@ -3782,8 +3805,8 @@
-                    <p>If <var>t</var>'s <a data-for=
-                    "RTCRtpTransceiver">direction</a> is "sendrecv" or
+                    <p>If <var>t</var>'s [[<a href=
+                    "#direction-slot">Direction</a>]] slot is "sendrecv" or
```
Let's do the "inactive" stuff in a separate issue/PR.

"sendrecv" is a string, but it's also an enum value. We should decide on syntax for enum values and use it everywhere.

adam-be replied:
Created issue #1370 to track the question about code-tags or strings for enum values.

taylor-b replied:
`inactive` doesn't need explicit handling here. This step is just to make sure an "a=msid" line gets added. If you set the direction to "inactive" and it was previously something else, one of the next two bullets will return `true`:

>- If connection's currentLocalDescription if of type "offer", and the direction of the associated m= section in neither the offer nor answer matches t's direction, return "true".
>- If connection's currentLocalDescription if of type "answer", and the direction of the associated m= section in the answer does not match t's direction intersected with the offered direction (as described in [JSEP] (section 5.3.1.)), return "true".

jan-ivar replied:
Makes sense, thanks! My only remaining issue here then is https://github.com/w3c/webrtc-pc/issues/1403.

adam-be commented on webrtc.html:
```diff
@@ -3794,15 +3817,15 @@
                     "RTCPeerConnection">currentLocalDescription</a></code>
                     if of type "offer", and the direction of the associated m=
                     section in neither the offer nor answer matches
-                    <var>t</var>'s <a data-for=
```
yes

adam-be commented on webrtc.html:
```diff
@@ -4961,10 +4984,11 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
+                      precisely, the [[<a href=
+                      "#curr-direction-slot">CurrentDirection</a>]] slot of the
+                      <code><a>RTCRtpTransceiver</a></code> associated with the
+                      sender has never had  a value of <code>sendrecv</code> or
```
fixed

stefhak commented:
I think we should merge now. @jan-ivar did you open all three issues talked about in https://github.com/w3c/webrtc-pc/pull/1321#pullrequestreview-43572624?

jan-ivar commented:
I haven't. @taylor-b would you mind opening them? I need to step away for a bit.

taylor-b commented:
Filed issues, though I didn't file one for the "inactive" thing, since I'm fairly confident it's not an issue. See comment: https://github.com/w3c/webrtc-pc/pull/1321/files#r122254074
